txpipe/calibrate.py

```diff
@@ -0,0 +1,151 @@
+
+        # We parallelize by bin.  This isn't ideal but we don't know the number
+        # of objects in each bin per chunk, so we can't parallelize in full.  This
+        #  is a quick stage though.
+        bins = list(range(nbin)) + ["all"]
+        my_bins = list(self.split_tasks_by_rank(bins))
+
+        # Print out which bins this proc will do, also as a prompt to the user
+        #  in case they're wondering why adding procs doesn't help
+        if my_bins:
+            my_bins_text = ", ".join(str(x) for x in my_bins)
+            print(f"Process {self.rank} collating bins: [{my_bins_text}]")
+        else:
+            print(f"Note: Process {self.rank} will not do anything.")
+
+        # make the iterator that loops through data
+        it = self.combined_iterators(
+            self.config["chunk_rows"],
+            # first file
+            "shear_catalog",
+            "shear",
+            cat_cols,
+            "shear_tomography_catalog",
+            "tomography",
+            ["source_bin"],
+            parallel=False,
+        )
+
+        #  Main loop
+        for s, e, data in it:
+
+            if self.rank == 0:
+                print(f"Rank 0 processing data {s:,} - {e:,}")
+            # Rename mcal_g1 -> g1 etc
+            self.rename_metacal(data)
+
+            #  Now output the calibrated bin data for this processor
+            for b in my_bins:
+
+                # Select objects to go in this bin
+                if b == "all":
+                    # the 2D case is any object from any other bin
+                    w = np.where(data["source_bin"] >= 0)
+                    cal = cal2d
+                else:
+                    # otherwise just objects in this bin
+                    w = np.where(data["source_bin"] == b)
+                    cal = cals[b]
+
+                # Cut down the data to just this selection for output
+                d = {name: data[name][w] for name in output_cols}
+
+                # Calibrate the shear columns
+                d["g1"], d["g2"] = cal.apply(d["g1"], d["g2"], subtract_mean=subtract_mean_shear)
+
+                # Write output, keeping track of sizes
+                splitter.write_bin(d, b)
+
+        splitter.finish(my_bins)
+        output_file.close()
+
+    def setup_output(self):
+        # count the expected number of objects per bin from the tomo data
+        with self.open_input("shear_tomography_catalog") as f:
+            counts = f["tomography/source_counts"][:]
+            count2d = f["tomography/source_counts_2d"][0]
+            nbin = len(counts)
+
+        # Prepare the calibrated output catalog
+        f = self.open_output("calibrated_shear_catalog", parallel=True)
+
+        #  we only retain these columns
+        cols = ["ra", "dec", "weight", "g1", "g2"]
+
+        # structure is /shear/bin_1, /shear/bin_2, etc
+        g = f.create_group("shear")
+        g.attrs["nbin"] = nbin
+        g.attrs["nbin_source"] = nbin
```

Comment:
unrelated... what's the difference between nbin and nbin_source?

Reply:
added a comment